Perforce filenames 2746

cross opened the pull request:
<!-- OCA submitted April 29, 2019.  -->

This avoids the errors from perforce when using filenames containing reserved characters, and will fix issue #2746 .

I am not a java developer so much, so I'm happy to adjust the code to match core opinions of how it should be implemented.  Let me know.

vladak commented on opengrok-indexer/src/test/java/org/opengrok/indexer/history/PerforceRepositoryTest.java:
```diff
@@ -35,8 +35,9 @@
+import java.util.AbstractMap.SimpleImmutableEntry;
 
-import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.*;
```
pls avoid wildcard imports

vladak commented on opengrok-indexer/src/main/java/org/opengrok/indexer/history/PerforceRepository.java:
```diff
@@ -61,14 +62,44 @@ public PerforceRepository() {
+        t = t.replaceAll("@", "%40");
+*/
+        if (name != t) {
+            LOGGER.log(Level.INFO,
```
this should probably be `Level.FINEST`

cross replied:
Yeah, that was originally for my debugging and I thought it might come out.  But leaving it in at FINE or FINEST makes sense, so I'll change it as you suggest.

vladak commented on opengrok-indexer/src/main/java/org/opengrok/indexer/history/PerforceRepository.java:
```diff
@@ -61,14 +62,44 @@ public PerforceRepository() {
+        for (SimpleImmutableEntry<String,String> ent : transmap) {
+            t = t.replaceAll(ent.getKey(), ent.getValue());
+        }
+        /* Instead of iterating over a list, just force each? */
```
not a bad idea at all. keep it simple.

cross replied:
So you'd prefer I just call replaceAll four times?  Easy enough, that's why I left the comment in.  I'll use that.

vladak commented on opengrok-indexer/src/main/java/org/opengrok/indexer/history/PerforceRepository.java:
```diff
@@ -61,14 +62,44 @@ public PerforceRepository() {
         ignoredFiles.add(".p4config");
     }
 
+    public static String protectPerforceFilename(String name) {
```
this method could be protected perhaps or package private

cross replied:
As it's called from PerforceHistoryParser, I assumed it needed to be public.  Is there a relationship I'm missing?

vladak replied:
The method can be package private (i.e. omit all the specifiers so the method signature becomes just `static String protectPerforceFilename(String name)`) because it is used by the class itself and also by the test case which resides in the same package.

If you use any decent IDE for editing the sources, it will suggest the change in fact.

cross replied:
Well, I'm old school, my IDE is vi.  :-)  Thanks, change made.

vladak commented on opengrok-indexer/src/test/java/org/opengrok/indexer/history/PerforceRepositoryTest.java:
```diff
@@ -105,4 +106,25 @@ public void testHistoryAndAnnotations() throws Exception {
+            String prot = instance.protectPerforceFilename(ent.getKey());
+            assertEquals("Improper protected filename, "+prot+" != "+ent.getValue(), ent.getValue(), prot);
+        }
+//        instance.update();
```
?

cross replied:
removed.  I think I didn't know why it was there.  And, I've since wondered if I should bother having an instance in this test, since it's testing a static method anyway.  Left a comment to that affect in the updated commit.

vladak commented on opengrok-indexer/src/test/java/org/opengrok/indexer/history/PerforceRepositoryTest.java:
```diff
@@ -105,4 +106,25 @@ public void testHistoryAndAnnotations() throws Exception {
+    @Test
+    public void testProtectFilename() throws Exception {
+        PerforceRepository instance = new PerforceRepository();
+        ArrayList<SimpleImmutableEntry<String,String>>   testmap = new ArrayList<SimpleImmutableEntry<String,String>>() {
```
I am pretty sure the diamond operator does not have to be fully specified on both sides.

cross replied:
I could not get it to compile without the inner types, I tried numerous omissions.  I am not very good at java, so may be wrong, but I didn't see how to make it work.

vladak replied:
Oh, this cannot be done because you are using anonymous inner class for initializing the array list.

The alternative I had in mind:
```java
        ArrayList<SimpleImmutableEntry<String,String>> testmap = new ArrayList<>();
        testmap.add(new SimpleImmutableEntry<>("Testfile 34", "Testfile 34"));
        testmap.add(new SimpleImmutableEntry<>("Test%52", "Test%2552"));
...
```

cross replied:
Ahh, okay.  I tried to be fancy by initializing in the declaration.  But, I see the value in the simpler approach you have above.  I'll do that, thanks.

cross replied:
Related question.  I have an instance of the class instantiated here, but am only testing a static method.  Would it be preferable just to call the static method, or instantiate a class object in this test?

cross replied:
Actually, before I push up my last commit.  @vladak - How would you prefer I engage this method in this test?  From an instanced instance of PerforceRepository, or just as a static method in PerforceRepository?  Neither is more correct, I don't think, so I leave it to preference.  Let me know, so I can take out the question comment here on that subject.

vladak replied:
Just use the static method directly.

vladak commented on opengrok-indexer/src/test/java/org/opengrok/indexer/history/PerforceRepositoryTest.java:
```diff
@@ -105,4 +106,25 @@ public void testHistoryAndAnnotations() throws Exception {
+        ArrayList<SimpleImmutableEntry<String,String>>   testmap = new ArrayList<SimpleImmutableEntry<String,String>>() {
+            {
+                add(new SimpleImmutableEntry<>("Testfile 34", "Testfile 34"));
+                add(new SimpleImmutableEntry<>("Test%52", "Test%2552"));
```
maybe add another test case with multiple occurrences of chars to be replaced (tests `relplaceAll()`)

vladak commented on opengrok-indexer/src/main/java/org/opengrok/indexer/history/PerforceHistoryParser.java:
```diff
@@ -87,7 +87,7 @@ public static History getRevisions(File file, String rev) throws IOException {
         cmd.add("filelog");
         cmd.add("-lti");
-        cmd.add(file.getName() + PerforceRepository.getRevisionCmd(rev));
+        cmd.add(PerforceRepository.protectPerforceFilename(file.getName()) + PerforceRepository.getRevisionCmd(rev));
```
import and use the method directly

cross replied:
I'm sorry, I don't understand what you're asking for here.  This is modeled after the other method calls to methods of PerforceRepository, including PerforceRepository.getRevisionCmd in this same line.  Can you clarify what you'd like to see?

vladak replied:
If you add
```java
import static org.opengrok.indexer.history.PerforceRepository.protectPerforceFilename;
```
then you can replace the line with just:
```java
            String prot = protectPerforceFilename(ent.getKey());
```

vladak replied:
same thing as with the test method - import the static method and use it directly without having to specify the class.

cross replied:
Okay.  Can do.  I was just following the example of the calls to `getRevisionCmd`, but happy to do it this way instead.  Thanks for explaining how to accomplish it.

cross commented:
I see that coveralls notes 0 of 17 new lines covered, but the unit test I added should cause coverage of the new method I added.  Do I have to do something to get that new method "covered"?

cross commented:
@vladak Can you look over my most recent commit which fixes some things you mentioned?  There are still other questions I have open in comments here, if I could get feedback we can close any open issues.  Thanks.

vladak commented:
One more note - you might want to add your copyright to the modified files.

vladak commented:
Also, if you add the text `fixes #2746` as a commit message, it will automatically close the related issue upon merging the pull request.

vladak commented:
I think the lack of coverage is caused by not running any Perforce test as part of running the build within Travis (which happens to be the environment used to submit the results to Coveralls). This is because the test class is annotated with `@ConditionalRun(RepositoryInstalled.PerforceInstalled.class)` so if Perforce runtime is not installed (which is the case with Travis) then all the tests in the test class will be skipped.

cross commented:
@vladak Do you have a suggestion as to whether I call the static method directly from the test function, or within an instance of PerforceRepository?  If you don't care, I'll leave it as it, but will remove the questioning comment before asking the PR be merged.

vladak commented:
Just do the static import and call it directly.

Dne pá 10. 5. 2019 21:03 uživatel Chris Ross <notifications@github.com>
napsal:

> @vladak <https://github.com/vladak> Do you have a suggestion as to
> whether I call the static method directly from the test function, or within
> an instance of PerforceRepository? If you don't care, I'll leave it as it,
> but will remove the questioning comment before asking the PR be merged.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oracle/opengrok/pull/2762#issuecomment-491397955>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAWMMDEBOAR3DBT7HSNMVMTPUXBJXANCNFSM4HJO7QCQ>
> .
>

cross commented:
Done and done.  Thanks @vladak .  Ready for merge if you agree.

vladak commented:
Looks fine, thanks !
